src/librbd/mirror/snapshot/RemoveGroupSnapshotRequest.cc

```diff
@@ -0,0 +1,164 @@
+
+  C_Gather *gather_ctx = new C_Gather(m_cct, ctx);
+  for (auto &snap : m_group_snap->snaps) {
+    if (snap.snap_id == CEPH_NOSNAP) {
+      continue;
+    }
+    ImageCtx *ictx = nullptr;
+    if (m_image_ctx_map.find(snap.image_id) != m_image_ctx_map.end()) {
+      ictx = m_image_ctx_map[snap.image_id];
+    }
+
+    if (!ictx) {
+      ldout(m_cct, 10) << "failed to remove individual snapshot" << dendl;
+      continue;
+    }
+
+    ldout(m_cct, 10) << "removing individual snapshot: "
+                     << snap.snap_id << ", from image id:"
+                     << snap.image_id << dendl;
+    remove_image_snapshot(ictx, snap.snap_id, gather_ctx);
+  }
+
+  gather_ctx->activate();
+
+}
+
+template <typename I>
+void RemoveGroupSnapshotRequest<I>::handle_remove_group_image_snapshots(int r) {
+  ldout(m_cct, 10) << "r=" << r << dendl;
+
+  if (r < 0) {
+    lderr(m_cct) << "failed to remove individual group image snapshots: "
+                 << cpp_strerror(r) << dendl;
+    finish(r);
+    return;
+  }
+
+  remove_group_snapshot();
+}
+
+template <typename I>
+void RemoveGroupSnapshotRequest<I>::remove_group_snapshot() {
+  ldout(m_cct, 10) << dendl;
+
+  librados::ObjectWriteOperation op;
+  cls_client::group_snap_remove(&op, m_group_snap->id);
+
+  auto aio_comp = create_rados_callback<
+    RemoveGroupSnapshotRequest<I>,
+    &RemoveGroupSnapshotRequest<I>::handle_remove_group_snapshot>(this);
+  int r = m_group_io_ctx.aio_operate(librbd::util::group_header_name(m_group_id),
+                                     aio_comp, &op);
+  ceph_assert(r == 0);
+  aio_comp->release();
+}
+
+template <typename I>
+void RemoveGroupSnapshotRequest<I>::handle_remove_group_snapshot(int r) {
+  ldout(m_cct, 10) << "r=" << r << dendl;
+
+  if (r < 0) {
+    lderr(m_cct) << "failed to remove group snapshot metadata: "
+                 << cpp_strerror(r) << dendl;
+    finish(r);
+    return;
+  }
+
+  finish(0);
+}
+
+template <typename I>
+void RemoveGroupSnapshotRequest<I>::remove_image_snapshot(ImageCtx *image_ctx,
+                                                      uint64_t snap_id,
+                                                      C_Gather *gather_ctx) {
+  ldout(m_cct, 10) << snap_id << dendl;
+
+  int r = -ENOENT;
+  cls::rbd::SnapshotNamespace snap_namespace;
+  std::string snap_name;
+
+  image_ctx->image_lock.lock_shared();
+  auto snap_it = image_ctx->snap_info.find(snap_id);
+  if (snap_it != image_ctx->snap_info.end()) {
+    r = 0;
+    snap_namespace = snap_it->second.snap_namespace;
+    snap_name = snap_it->second.name;
+  }
+
+  image_ctx->image_lock.unlock_shared();
+  if (r == -ENOENT) {
+    ldout(m_cct, 10) << "missing snapshot: snap_id=" << snap_id << dendl;
+    return;
+  }
+  auto mirror_ns = std::get_if<cls::rbd::MirrorSnapshotNamespace>(
+    &snap_namespace);
+  if (mirror_ns == nullptr) {
+    lderr(m_cct) << "not mirror snapshot (snap_id=" << snap_id << ")" << dendl;
```

Comment:
Not sure how this can happen: a  group image snapshot of a mirror group snapshot belongs to a non-mirror snapshot namespace?